pydantic_ai_slim/pydantic_ai/models/xai.py

```diff
@@ -0,0 +1,1046 @@
+
+        Returns:
+            A tuple of (instructions, xai_messages) where instructions may be None.
+        """
+        xai_messages: list[chat_types.chat_pb2.Message] = []
+
+        for message in messages:
+            if isinstance(message, ModelRequest):
+                xai_messages.extend(await self._map_request_parts(message.parts))
+            elif isinstance(message, ModelResponse):
+                if response_msg := self._map_response_parts(message.parts):
+                    xai_messages.append(response_msg)
+            else:
+                assert_never(message)
+
+        instructions = self._get_instructions(messages, model_request_parameters)
+        return instructions, xai_messages
+
+    async def _map_request_parts(self, parts: Sequence[ModelRequestPart]) -> list[chat_types.chat_pb2.Message]:
+        """Map ModelRequest parts to xAI messages."""
+        xai_messages: list[chat_types.chat_pb2.Message] = []
+
+        for part in parts:
+            if isinstance(part, SystemPromptPart):
+                xai_messages.append(system(part.content))
+            elif isinstance(part, UserPromptPart):
+                if user_msg := await self._map_user_prompt(part):
+                    xai_messages.append(user_msg)
+            elif isinstance(part, ToolReturnPart):
+                xai_messages.append(tool_result(part.model_response_str()))
+            elif isinstance(part, RetryPromptPart):
+                if part.tool_name is None:
+                    # Retry prompt as user message
+                    xai_messages.append(user(part.model_response()))
+                else:
+                    # Retry prompt as tool result
+                    xai_messages.append(tool_result(part.model_response()))
+            else:
+                assert_never(part)
+
+        return xai_messages
+
+    def _map_response_parts(self, parts: Sequence[ModelResponsePart]) -> chat_types.chat_pb2.Message | None:
+        """Map ModelResponse parts to an xAI assistant message."""
+        # Collect content from response parts
+        texts: list[str] = []
+        reasoning_contents: list[str] = []
+        encrypted_contents: list[str] = []
+        tool_calls: list[chat_types.chat_pb2.ToolCall] = []
+
+        # Track builtin tool calls to update their status with corresponding return parts
+        code_execution_tool_call: chat_types.chat_pb2.ToolCall | None = None
+        web_search_tool_call: chat_types.chat_pb2.ToolCall | None = None
+        mcp_tool_call: chat_types.chat_pb2.ToolCall | None = None
+
+        for item in parts:
+            if isinstance(item, TextPart):
+                texts.append(item.content)
+            elif isinstance(item, ThinkingPart):
+                # xAI models (grok) support reasoning_content directly
+                if item.content:
+                    reasoning_contents.append(item.content)
+                if item.signature:
```

Comment:
We should only do this if `ThinkingPart.provider_name` matches `self.system`, since it's possible to send history from one provider to another provider, and we don't want the provider to complain about encrypted reasoning it can't decrypt